File: lib/services/inquiries.js
import db from '../database.js';

/**
 * Inquiries Service
 * Handles all inquiry-related database operations
 */

export class InquiriesService {
  constructor() {
    this.filename = 'leads.json';
  }

  /**
   * Get all inquiries
   */
  async getAll() {
    try {
      const inquiries = await db.readData(this.filename, []);
      return inquiries.sort((a, b) => 
        new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime()
      );
    } catch (error) {
      console.error('Error getting all inquiries:', error);
      return [];
    }
  }

  /**
   * Get inquiry by ID
   */
  async getById(id) {
    try {
      const inquiries = await this.getAll();
      return inquiries.find(inquiry => inquiry.id === id) || null;
    } catch (error) {
      console.error('Error getting inquiry by ID:', error);
      return null;
    }
  }

  /**
   * Create new inquiry
   */
  async create(inquiryData) {
    try {
      const inquiries = await this.getAll();
      
      const newInquiry = {
        id: db.generateId('inquiry'),
        createdAt: db.getTimestamp(),
        status: 'new',
        ...inquiryData
      };

      inquiries.push(newInquiry);
      await db.writeData(this.filename, inquiries);
      
      console.log('✅ Inquiry created successfully:', newInquiry.id);
      return newInquiry;
    } catch (error) {
      console.error('Error creating inquiry:', error);
      throw new Error('Failed to create inquiry');
    }
  }

  /**
   * Update inquiry
   */
  async update(id, updates) {
    try {
      const inquiries = await this.getAll();
      const index = inquiries.findIndex(inquiry => inquiry.id === id);
      
      if (index === -1) {
        throw new Error('Inquiry not found');
      }

      // Update status-specific timestamps
      if (updates.status && ['accepted', 'declined'].includes(updates.status)) {
        updates.reviewedAt = db.getTimestamp();
      }

      inquiries[index] = { 
        ...inquiries[index], 
        ...updates,
        updatedAt: db.getTimestamp()
      };

      await db.writeData(this.filename, inquiries);
      
      console.log('✅ Inquiry updated successfully:', id);
      return inquiries[index];
    } catch (error) {
      console.error('Error updating inquiry:', error);
      throw new Error('Failed to update inquiry');
    }
  }

  /**
   * Delete inquiry
   */
  async delete(id) {
    try {
      const inquiries = await this.getAll();
      const filteredInquiries = inquiries.filter(inquiry => inquiry.id !== id);
      
      if (filteredInquiries.length === inquiries.length) {
        throw new Error('Inquiry not found');
      }

      await db.writeData(this.filename, filteredInquiries);
      
      console.log('✅ Inquiry deleted successfully:', id);
      return true;
    } catch (error) {
      console.error('Error deleting inquiry:', error);
      throw new Error('Failed to delete inquiry');
    }
  }

  /**
   * Get inquiries by status
   */
  async getByStatus(status) {
    try {
      const inquiries = await this.getAll();
      return inquiries.filter(inquiry => inquiry.status === status);
    } catch (error) {
      console.error('Error getting inquiries by status:', error);
      return [];
    }
  }

  /**
   * Get inquiry statistics
   */
  async getStats() {
    try {
      const inquiries = await this.getAll();
      
      return {
        total: inquiries.length,
        new: inquiries.filter(inquiry => inquiry.status === 'new').length,
        accepted: inquiries.filter(inquiry => inquiry.status === 'accepted').length,
        declined: inquiries.filter(inquiry => inquiry.status === 'declined').length
      };
    } catch (error) {
      console.error('Error getting inquiry stats:', error);
      return { total: 0, new: 0, accepted: 0, declined: 0 };
    }
  }
}

// Create and export a single instance
const inquiriesService = new InquiriesService();
export default inquiriesService;

File: app/services/inquiries.py
from typing import List, Dict, Any, Optional
from .database import db

class InquiriesService:
    def __init__(self):
        self.filename = 'leads.json'
    
    async def get_all(self) -> List[Dict[str, Any]]:
        """Get all inquiries sorted by creation date"""
        try:
            inquiries = await db.read_data(self.filename, [])
            return sorted(inquiries, key=lambda x: x.get('createdAt', ''), reverse=True)
        except Exception as e:
            print(f"Error getting all inquiries: {e}")
            return []
    
    async def get_by_id(self, inquiry_id: str) -> Optional[Dict[str, Any]]:
        """Get inquiry by ID"""
        try:
            inquiries = await self.get_all()
            return next((inq for inq in inquiries if inq.get('id') == inquiry_id), None)
        except Exception as e:
            print(f"Error getting inquiry by ID: {e}")
            return None
    
    async def create(self, inquiry_data: Dict[str, Any]) -> Dict[str, Any]:
        """Create new inquiry"""
        try:
            inquiries = await self.get_all()
            
            new_inquiry = {
                'id': db.generate_id('inquiry'),
                'createdAt': db.get_timestamp(),
                'status': 'new',
                **inquiry_data
            }
            
            inquiries.append(new_inquiry)
            await db.write_data(self.filename, inquiries)
            
            print(f"✅ Inquiry created successfully: {new_inquiry['id']}")
            return new_inquiry
        except Exception as e:
            print(f"Error creating inquiry: {e}")
            raise Exception("Failed to create inquiry")
    
    async def update(self, inquiry_id: str, updates: Dict[str, Any]) -> Dict[str, Any]:
        """Update inquiry"""
        try:
            inquiries = await self.get_all()
            index = next((i for i, inq in enumerate(inquiries) if inq.get('id') == inquiry_id), -1)
            
            if index == -1:
                raise Exception("Inquiry not found")
            
            # Update status-specific timestamps
            if updates.get('status') in ['accepted', 'declined']:
                updates['reviewedAt'] = db.get_timestamp()
            
            inquiries[index] = {
                **inquiries[index],
                **updates,
                'updatedAt': db.get_timestamp()
            }
            
            await db.write_data(self.filename, inquiries)
            print(f"✅ Inquiry updated successfully: {inquiry_id}")
            return inquiries[index]
        except Exception as e:
            print(f"Error updating inquiry: {e}")
            raise Exception("Failed to update inquiry")
    
    async def delete(self, inquiry_id: str) -> bool:
        """Delete inquiry"""
        try:
            inquiries = await self.get_all()
            filtered_inquiries = [inq for inq in inquiries if inq.get('id') != inquiry_id]
            
            if len(filtered_inquiries) == len(inquiries):
                raise Exception("Inquiry not found")
            
            await db.write_data(self.filename, filtered_inquiries)
            print(f"✅ Inquiry deleted successfully: {inquiry_id}")
            return True
        except Exception as e:
            print(f"Error deleting inquiry: {e}")
            raise Exception("Failed to delete inquiry")
    
    async def get_by_status(self, status: str) -> List[Dict[str, Any]]:
        """Get inquiries by status"""
        try:
            inquiries = await self.get_all()
            return [inq for inq in inquiries if inq.get('status') == status]
        except Exception as e:
            print(f"Error getting inquiries by status: {e}")
            return []
    
    async def get_stats(self) -> Dict[str, int]:
        """Get inquiry statistics"""
        try:
            inquiries = await self.get_all()
            return {
                'total': len(inquiries),
                'new': len([inq for inq in inquiries if inq.get('status') == 'new']),
                'accepted': len([inq for inq in inquiries if inq.get('status') == 'accepted']),
                'declined': len([inq for inq in inquiries if inq.get('status') == 'declined'])
            }
        except Exception as e:
            print(f"Error getting inquiry stats: {e}")
            return {'total': 0, 'new': 0, 'accepted': 0, 'declined': 0}

# Global service instance
inquiries_service = InquiriesService()
